core/src/main/java/io/parsingdata/metal/token/IterableToken.java

```diff
@@ -0,0 +1,64 @@
+package io.parsingdata.metal.token;
+
+import io.parsingdata.metal.Trampoline;
+import io.parsingdata.metal.data.Environment;
+import io.parsingdata.metal.data.ParseState;
+import io.parsingdata.metal.encoding.Encoding;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static io.parsingdata.metal.Trampoline.complete;
+import static io.parsingdata.metal.Trampoline.intermediate;
+import static io.parsingdata.metal.Util.checkNotNull;
+import static io.parsingdata.metal.Util.success;
+
+public abstract class IterableToken extends Token {
```

Comment:
I am usually a bit skeptical of abstract base classes, but this looks like a good use of one. This has deduplicated quite a bit of complex code, and kept the most important concerns in the implementations: what is the stop condition, and what do we do when parsing fails? I like! 👍